src/algorithms/base_algorithm.cpp

```diff
@@ -1,10 +1,60 @@
 #include "base_algorithm.hpp"
 
 #include <Eigen/Dense>
+#include <cassert>
 
 #include "../../proto/cpp/marginal_state.pb.h"
 #include "../../proto/cpp/mixings.pb.h"
 
+void BaseAlgorithm::initialize() {
+  std::cout << "Initializing..." << std::endl;
+
+  // Perform checks
+  assert(data.rows() != 0 && "Error: empty data matrix");
+  assert(unique_values.size() != 0 && "Error: hierarchy was not provided");
+  assert(!(unique_values[0]->is_multivariate() == false && data.cols() > 1) &&
+         "Error: multivariate data supplied to univariate hierarchy");
+
+  if (init_num_clusters == 0) {
+    init_num_clusters = data.rows();
+  }
+
+  // Initialize hierarchies
+  unique_values[0]->initialize();
+  for (size_t i = 0; i < init_num_clusters - 1; i++) {
+    unique_values.push_back(unique_values[0]->clone());
+  }
+
+  // Initialize needed objects
+  cardinalities.clear();
+  cardinalities.reserve(data.rows());
+  std::default_random_engine generator;
+  // Build uniform probability on clusters, given their initial number
+  std::uniform_int_distribution<int> distro(0, init_num_clusters - 1);
+  // Allocate one datum per cluster first, and update cardinalities
+  allocations.clear();
+  for (size_t i = 0; i < init_num_clusters; i++) {
+    allocations.push_back(i);
+    cardinalities.push_back(1);
+  }
+  // Randomly allocate all remaining data, and update cardinalities
+  for (size_t i = init_num_clusters; i < data.rows(); i++) {
+    unsigned int clust = distro(generator);
+    allocations.push_back(clust);
+    cardinalities[clust] += 1;
+  }
+}
+
+void BaseAlgorithm::update_hierarchy_hypers() {
+  bayesmix::MarginalState::ClusterVal clust;
```

Comment:
I don't follow. Neal8 auxiliary blocks are already stored in a separate vector, aux_unique_values, therefore they are unaffected by update_hierarchy_hypers(). Should we change anything?

Reply:
Yeah you're right, my bad. 
Still please add a reminder here, because when we move to the ConditionalAlgorithm (like logit stick breaking), this would still run but would be wrong.
We can change stuff when we get to that point but it's going to be hard to keep track of all these without some reminders